pallets/randomness-beacon/src/lib.rs

```diff
@@ -283,28 +281,27 @@ pub mod pallet {
 			round: Option<RoundNumber>,
 		) -> DispatchResult {
 			ensure_none(origin)?;
-			ensure!(!DidUpdate::<T>::exists(), Error::<T>::SignatureAlreadyVerified,);
-
-			let config = T::BeaconConfig::get();
-			let mut genesis_round = GenesisRound::<T>::get();
-			let mut latest_round = LatestRound::<T>::get();
-
+			// the extrinsic can only be successfully executed once per block
+			ensure!(!DidUpdate::<T>::exists(), Error::<T>::SignatureAlreadyVerified);
+			// ensure the height is within [1, T::MaxSigsPerBlock]
 			ensure!(height > 0, Error::<T>::ZeroHeightProvided);
 			ensure!(
 				height <= T::MaxSigsPerBlock::get() as u64,
 				Error::<T>::ExcessiveHeightProvided
 			);
 
-			if let Some(r) = round {
-				// if a round is provided and the genesis round is not set
-				ensure!(genesis_round == 0, Error::<T>::GenesisRoundAlreadySet);
-				GenesisRound::<T>::set(r);
-				genesis_round = r;
-				latest_round = genesis_round;
-			} else {
-				//  if the genesis round is not set and a round is not provided
-				ensure!(GenesisRound::<T>::get() > 0, Error::<T>::GenesisRoundNotSet);
-			}
+			let config = T::BeaconConfig::get();
+			// if the genesis round is not configured, do it on the first pass through
+			let genesis_round: RoundNumber = GenesisRound::<T>::get().unwrap_or_else(|| {
+				// return 0 if round is not provided and genesis round not set
+				let r = round.unwrap_or(0);
+				GenesisRound::<T>::set(round);
+				r
+			});
+
+			ensure!(genesis_round > 0, Error::<T>::GenesisRoundNotSet);
```

Comment:
I would avoid treating `0` as the absence of value, let's stick to `None`. Also, we should immediately error out if both `GenesisRound` and `round` are `None` there's no need to continue computing anything

```suggestion
			let genesis_round: RoundNumber = match GenesisRound::<T>::get() {
				Some(gr) => gr,
				None => {
					// error if the genesis round is not set and a round is not provided
					let r = round.ok_or(Error::<T>::GenesisRoundNotSet)?;
					GenesisRound::<T>::set(round);
					r
				},
			};
```

Reply:
Updated!